### PR TITLE
Fix gradient in tape mode when measurements are independent of a parameter

### PR DESCRIPTION
**Pull request by trbromley**
This PR fixes a bug in tape mode that occurs when some parameters are independent of a measurement. Consider the following snippet:
```python
import pennylane as qml
from pennylane import numpy as np
qml.enable_tape()

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circ(params):
    qml.RX(params[0], wires=0)
    qml.RX(params[1], wires=1)
    return qml.expval(qml.PauliZ(0))

params = np.ones(2)

qml.grad(circ)(params)
```
This takes the gradient of a PauliZ expectation value on the first qubit. We expect to get `(array([-0.52167534, 0]),)` since the second parameter has no effect on the first qubit. Instead, we incorrectly get `(array([-0.52167534, -0.52167534]),)`.

Now suppose we have `return qml.expval(qml.PauliZ(1))`. Instead of getting an incorrect result, we get a `UnboundLocalError: local variable 'processing_fn' referenced before assignment` error.

Both of these problems occur because the `processing_fn` is not defined if `param_method == "0"`, and so we either use a previously defined `processing_fn` or raise an error.

This PR fixes the above problem by skipping the cases where `param_method == "0"`.

### CODE REVIEW

File: pennylane/tape/tapes/jacobian_tape.py [hunk 1]
```diff
@@ -490,24 +490,28 @@ def jacobian(self, device, params=None, **options):
...
-            all_tapes.extend(tapes)
+                processing_fns.append(processing_fn)
 
-            # to extract the correct result for this parameter later, remember the number of tapes
-            reshape_info.append(len(tapes))
+                # we create a flat list here to feed at once to the device
+                all_tapes.extend(tapes)
+
+                # to extract the correct result for this parameter later, remember the number of tapes
+                reshape_info.append(len(tapes))
```
Comment (trbromley) [inline pennylane/tape/tapes/jacobian_tape.py:514]: This is all just indenting what we had before

**Comment by trbromley** [comment]
[ch2468]

File: pennylane/tape/tapes/jacobian_tape.py [hunk 2]
```diff
@@ -490,24 +490,28 @@ def jacobian(self, device, params=None, **options):
...
         processing_fns = []
+        nonzero_grad_idx = []
 
         for trainable_idx, param_method in enumerate(diff_methods):
+            if param_method != "0":
 
-            if (method == "best" and param_method[0] == "F") or (method == "numeric"):
-                # numeric method
-                tapes, processing_fn = self.numeric_pd(trainable_idx, params=params, **options)
+                nonzero_grad_idx.append(trainable_idx)
```
Comment (josh146) [inline pennylane/tape/tapes/jacobian_tape.py:498]: Minor suggestion: you could do the following,
```python
for trainable_idx, param_method in enumerate(diff_methods):
    if param_method == "0":
        continue

    nonzero_grad_idx.append(trainable_idx)

    if (method == "best" and param_method[0] == "F") or (method == "numeric"):
        # numeric method
        tapes, processing_fn = self.numeric_pd(trainable_idx, params=params, **options)

    elif (method == "best" and param_method[0] == "A") or (method == "analytic"):
        # analytic method
        tapes, processing_fn = self.analytic_pd(trainable_idx, params=params, **options)

    processing_fns.append(processing_fn)

    # we create a flat list here to feed at once to the device
    all_tapes.extend(tapes)

    # to extract the correct result for this parameter later, remember the number of tapes
    reshape_info.append(len(tapes))
```

While no change to the logic, it helps with readability (since it avoids the extra indent).

File: pennylane/tape/tapes/jacobian_tape.py [hunk 3]
```diff
@@ -516,7 +520,7 @@ def jacobian(self, device, params=None, **options):
         jac = None
         start = 0
 
-        for i, (processing_fn, res_len) in enumerate(zip(processing_fns, reshape_info)):
+        for i, processing_fn, res_len in zip(nonzero_grad_idx, processing_fns, reshape_info):
```
Comment (josh146) [inline pennylane/tape/tapes/jacobian_tape.py:523]: 💪

File: tests/tape/tapes/test_jacobian_tape.py [hunk 4]
```diff
@@ -384,6 +384,33 @@ def test_numeric_unknown_order(self):
...
+
+        with JacobianTape() as tape1:
+            qml.RX(1, wires=[0])
+            qml.RX(1, wires=[1])
+            qml.expval(qml.PauliZ(0))
+
+        with JacobianTape() as tape2:
+            qml.RX(1, wires=[0])
+            qml.RX(1, wires=[1])
+            qml.expval(qml.PauliZ(1))
```
Comment (josh146) [inline tests/tape/tapes/test_jacobian_tape.py:400]: Awesome, this catches both edge cases in your PR comment 👍

**Review by josh146 — APPROVED** [review]
Thanks @trbromley! This was a nasty bug, and shows we had holes in our existing test coverage

File: pennylane/tape/tapes/jacobian_tape.py (hunk 2 above)
Comment (trbromley) [reply pennylane/tape/tapes/jacobian_tape.py:498]: Thanks, done!
